Adding composition for aws aurora db & proxy (#137)
awslabs/crossplane-on-eks: examples/upbound-aws-provider/composite-resources/database-examples/aurora/README.md

--role-name aurora-monitoring \
--policy-arn arn:aws:iam::aws:policy/service-role/AmazonRDSEnhancedMonitoringRole
```

Create an 2nd IAM role and attach policy to the role (This role is required for aurora proxy to fetch db secrets from secret manager )

```shell
aws iam create-role \
--role-name rds-proxy \
--assume-role-policy-document file://rds-proxy.json
```

 ```shell
aws iam put-role-policy \
--role-name rds-proxy \
--policy-name rds-proxy-policy \
--policy-document file://rds-proxy-policy.json
```
  ### Provide value to the claim before applying.
     
 1. Open the claim and substitute aurora-monitoring role arn with monitoringRoleArn and rds-proxy role arn with proxyRoleArn respectively.
 2. Make sure check the CIDR block for application and provide in the claim. This will used as the security group 
    ingress rule for proxy from application ( we can also provide the security group id instead of CIDR).    
    
 ### Verify your XRDS and composition

  Verify the XRDs

```shell
kubectl get xrds | grep xauroras.db.awsblueprint.io
```

Expected output:

NAME                          ESTABLISHED   OFFERED   AGE
xauroras.db.awsblueprint.io   True          True      5m

> 

Reviewer: add shell

Author: shell added.

Verify the Compositions
    
 ```shell
kubectl get compositions | grep xauroras.db.awsblueprint.io 
```

Expected output:

NAME                          XR-KIND   XR-APIVERSION                 AGE
xauroras.db.awsblueprint.io   XAurora   db.awsblueprint.io/v1alpha1   5m

> 

Reviewer: add shell

Author: shell added and committed.

```
Create a service account passing the above policy arn

```shell
POLICY_ARN=$(aws iam list-policies --query 'Policies[?PolicyName==`rdsproxy-access`].Arn' --output text)
CLUSTER_NAME=$(aws eks list-clusters --query 'clusters[]' --output text | grep <YOUR-CLUSTER-NAME>)

eksctl create iamserviceaccount \
--name rds-access \
--namespace team-a \
--cluster ${CLUSTER_NAME} \
--attach-policy-arn ${POLICY_ARN} \
--approve \
-override-existing-serviceaccounts
``` 
Generate a DB token which can be used for password when asked

```shell
PROXY_TERGET-ENDPOINT=k get secrets aurora-cluster-secrets  -n team-a -o json | jq -r .data.proxyEndpoint | base64 --decode
CLUSTER_USER_NAME=k get secrets aurora-cluster-secrets  -n team-a -o json | jq -r .data.clusterUsername | base64 --decode

aws rds generate-db-auth-token \
--hostname ${PROXY_TERGET-ENDPOINT}  \
--port 5432 \
--region ${REGION} \
--username ${CLUSTER_USER_NAME}
```
create the pod with psql client:

```shell
k apply -f psql-client-pod.yaml
```

Exec into the pod

```shell
k exec -it postgres-client -n team-a  -- sh
```

create a DB connection and provide the token which you generate earlier for password.

```shell

> psql -h ${PROXY_TERGET-ENDPOINT}  -U ${CLUSTER_USER_NAME} -d aurorapgsqldb -W

Reviewer: replace variable names

Author: replaced the variable name as it would be available in the pod.

```
You should be able to connect to the db now then, lets check the list of default table 

```shell
\dt
```
Now we can do all the required database operation.
Note : we are providing the name of the database in the claim as:  aurorapgsqldb

### Deleting the resources:

First delete the psql client 
```shell
k delete -f psql-client-pod.yaml
```
Delete the PostgreSQL Database by deleting the claim
```shell
k delete -f aurora-postgresql.yaml
```
Note: It will take around ~15-20 min delete the whole cluster

Delete the composition and XRDs
```shell
cd ../composition/upbound-aws-provider/aurora
k delete -f aurora.yaml

> k delete -f defination.yaml

Reviewer: replace all k with kubectl

Author: done with this change.

```

> Note : The role aurora-monitoring & rds-proxy will eventually be part of composition , till then it will be required to create these roles external to the composition.

Reviewer: Move to top and include github issue.

Author: Created an issue and provided the reference. The comment is now placed on top.

# Steps to to deploy aurora rds cluster
This example deploys an Aurora PostgreSQL database cluster.
Below are the steps to create a cluster and connect from a pod with an psql client.

## Pre-requisites
 - [Upbound AWS Provider Crossplane Blueprint Examples](../../README.md)

> Note : aurora-monitoring & rds-proxy role will eventually be part of composition , till then we need to create that 2 role external to the composition.

Reviewer: ```suggestion
[!NOTE]\ : aurora-monitoring and rds-proxy iam roles need to created manually at this time but will be added to the composition, track progress in [this issue](https://github.com/awslabs/crossplane-on-eks/issues/144).
```

Author: updated the note as provided.

> Reference of the issue created for tracking: 
> - [Add Aurora monitoring and rds proxy role as part of the aurora composition](https://github.com/awslabs/crossplane-on-eks/issues/144)

Reviewer: ```suggestion
```

# Steps to to deploy aurora rds cluster

> This example deploys an Aurora PostgreSQL database cluster.

Reviewer 2: ```suggestion
This example deploys an Aurora PostgreSQL database cluster with Amazon RDS Proxy for Aurora.
```

Author: updated based on the comments.

### Create 2 roles, one for RDS Monitoring and one for RDS Proxy

Create an IAM role and attach policy to the role (This role is required for aurora to perform monitoring)

    
```shell
aws iam create-role \
--role-name aurora-monitoring \
--assume-role-policy-document file://aurora-monitoring.json
 ```
 (Attach the IAM role with AmazonRDSEnhancedMonitoringRole policy.)

```shell
aws iam attach-role-policy \
--role-name aurora-monitoring \
--policy-arn arn:aws:iam::aws:policy/service-role/AmazonRDSEnhancedMonitoringRole
```

Create second IAM role and attach policy to the role (This role is required for aurora proxy to fetch db secrets from secret manager )

```shell
aws iam create-role \
--role-name rds-proxy \
--assume-role-policy-document file://rds-proxy.json
```
 
 ```shell
CURRENT_REGION=<REGION_NAME> # replace <REGION_NAME> with current region name e.g us-east-1.
CLUSTER_NAME=<EKS_CLUSTER_NAME> # replace <EKS_CLUSTER_NAME> with eks cluster name.
ACCOUNT_NUM=$(aws sts get-caller-identity --query Account --output text)
```

```shell

> sed -i -e "s/REGION-NAME/$CURRENT_REGION/g" rds-proxy-policy.json

Reviewer 2: Let's use `envsubst` instead.

Author: updated based on the comment.

--policy-document file://rds-proxy-policy.json
```
### Provide value to the claim before applying.
     
 1. Open the claim and substitute aurora-monitoring role arn with monitoringRoleArn and rds-proxy role arn with proxyRoleArn respectively.
 2. Make sure check the CIDR block for application and provide in the claim. This will used as the security group 
    ingress rule for proxy from application ( we can also provide the security group id instead of CIDR).    
    
### Deploy and verify XRD and composition

Deploy the xrd and composition for aurora
```shell
# Assuming root directory.
cd ./compositions/upbound-aws-provider/aurora
kubectl apply -k .
```

```shell
kubectl get xrds | grep xauroras.db.awsblueprint.io
```

Expected output:

```shell
NAME                          ESTABLISHED   OFFERED   AGE
xauroras.db.awsblueprint.io   True          True      5m
```

Verify the Compositions
    
 ```shell
kubectl get compositions | grep xauroras.db.awsblueprint.io 
```

Expected output:

```shell
NAME                          XR-KIND   XR-APIVERSION                 AGE
xauroras.db.awsblueprint.io   XAurora   db.awsblueprint.io/v1alpha1   5m
```

### Apply the Aurora claim

> If we have about same kind of output as above we can apply th claim: ( We are using a namespace as team-a in this example, if you want to change , please go ahead and create and update the same in the claim)

Author: updated.

 ```shell
cd ../../../examples/upbound-aws-provider/composite-resources/database-examples/aurora
kubectl apply -f aurora-postgresql.yaml
```

We can check the execution of claim by following command:

```shell
kubectl get Aurora -n team-a
kubectl describe Aurora -n team-a
```

> (It should take about 15-20 min to provision the Aurora RDS cluster.)

Reviewer 2: Also provide how you can track the progress in K8s.

> (It should take about 15-20 min to provision the Aurora RDS cluster.)

Author: Added the k8s command.

> Below is the default behaviour of the resource which will be provisioned through the claim, just to mention all this default behaviour can be overridden through patching .

Reviewer 2: You state the default behaviour but you don't tell how it can be changed. You should at least point to fields that can be modified.

Author: Added the aws document url.

 1. It will create a 3 node cluster with 1 writer and 2 reader endpoint spread across 3 Azs by default.

>  2. It will also create a proxy to connect to the aurora cluster.

Reviewer 2: There are a lot of mentions about proxies in this doc but not seeing a link or description as to what it is. I think you should at least link to the Aurora doc somewhere.

Author: Added the AWS document url.

 3. The management of database credential is done through the secret manager.
 4. The security group has 2 rules , one to allow the app to connect to the proxy and one within the security group
     where the proxy can connect to the database (Make sure you have provided the correct the CIDR at the claim which is the allow CIDR on the security group from app.)
 5. The Aurora DB has been configured with logging and monitoring.

### Test - Steps to check the database connectivity from a pod
Create a  policy to allow the pod to connect to the Aurora Database
  
```shell
# provide the account number and region name to the policy document.

> sed -i -e "s/REGION-NAME:ACCOUNT-NUM/$CURRENT_REGION:$ACCOUNT_NUM/g" rdsproxy-access.json

Reviewer 2: Let's use `envsubst` instead.

Author: using envsubst now.

aws iam create-policy \
--policy-name rdsproxy-access \
--policy-document file://rdsproxy-access.json
```
Create a service account passing the above policy arn

```shell
POLICY_ARN=$(aws iam list-policies --query 'Policies[?PolicyName==`rdsproxy-access`].Arn' --output text)

eksctl create iamserviceaccount \
--name rds-access \
--namespace team-a \
--cluster ${CLUSTER_NAME} \
--attach-policy-arn ${POLICY_ARN} \
--approve \
-override-existing-serviceaccounts
``` 
Generate a DB token which can be used for password when asked

```shell

> PROXY_TERGET-ENDPOINT=kubectl get secrets aurora-cluster-secrets  -n team-a -o json | jq -r .data.proxyEndpoint | base64 --decode

Reviewer 2: ```suggestion
PROXY_TARGET_ENDPOINT=kubectl get secrets aurora-cluster-secrets  -n team-a -o json | jq -r '.data.proxyEndpoint' | base64 --decode
```

Author: updated based on the comments.

First delete the psql client pod
```shell
kubectl delete -f psql-client-pod.yaml
```
Delete the PostgreSQL Database by deleting the claim
```shell
kubectl delete -f aurora-postgresql.yaml
```
Note: It will take around ~15-20 min delete the whole cluster

Delete the Roles created
```shell
# Delete the serviceaccount
eksctl delete iamserviceaccount --name rds-access --cluster ${CLUSTER_NAME} --namespace team-a

# delete the policy created for serviceaccount
aws iam delete-policy --policy-arn $(aws iam list-policies --query 'Policies[?PolicyName==`rdsproxy-access`].Arn' --output text)

# Delete the rds-proxy-policy attached to the role
aws iam delete-role-policy --policy-name rds-proxy-policy --role-name rds-proxy

# Delete the rds-proxy role
aws iam delete-role --role-name rds-proxy

# Delete the rds-proxy-policy attached to the role
aws iam detach-role-policy \
--role-name aurora-monitoring \
--policy-arn arn:aws:iam::aws:policy/service-role/AmazonRDSEnhancedMonitoringRole

# Delete aurora-monitoring role
aws iam delete-role --role-name aurora-monitoring
``` 

Delete the composition and XRDs
```shell
cd ../composition/upbound-aws-provider/aurora
kubectl delete -f aurora.yaml

> kubectl delete -f defination.yaml

Reviewer 2: ```suggestion
kubectl delete -f definition.yaml
```

Author: fixed the typo.

# Steps to to deploy aurora rds cluster

> This example deploys an aurora postgresql database cluster when applied the claim. Below are the steps to execute to

Reviewer 2: ```suggestion
This example deploys an Aurora PostgreSQL database cluster. 
```

Author: Updated as mentioned.

> create a cluster and connect the cluster from a pod with an psql client.

Reviewer 2: ```suggestion
create a cluster and connect to the cluster from a pod with an psql client.
```

Author: updated accordingly.

## Pre-requisites
 - [Upbound AWS Provider Crossplane Blueprint Examples](../../README.md)

> ### create 2 role, one for RDS Monitoring and one for RDS Proxy

Author: updated accordingly.

> Create an IAM role and attach policy to the role (This role is required for aurora to perform monitoring)

Reviewer 2: Why does this need to be done outside of Crossplane?

Reviewer: Hi @nabuskey , I talked to @biswajitbabu and for the first pass we can have it in the Readme and then he will do another iteration to add it into the composition.

Reviewer 2: I see. In that case, let's at least put a disclaimer or something in readme and create an issue to track it. I suspect the first thing users are going to wonder when reading this is "why do we have to create roles manually?". Having an issue to track this will be great to invite others to fix this if it doesn't get done in the near future.

Author: The changes are done as discussed.

    
```shell

>     aws iam create-role --role-name aurora-monitoring \

Reviewer 2: ```suggestion
 aws iam create-role --role-name aurora-monitoring \
```

                            --assume-role-policy-document '{

>             "Version": "2012-10-17",

Reviewer 2: Could you please pretty print / format json?

Author: Updated as mentioned and moved the json to a different file in the same folder

            "Statement": [
            {
                "Sid": "",
                "Effect": "Allow",
                    "Principal":
                    {
                    "Service": "monitoring.rds.amazonaws.com"
                    },
                    "Action": "sts:AssumeRole"

            }
        ]
        }'
 ```
 (Attach the IAM role with AmazonRDSEnhancedMonitoringRole policy.)

```shell
    

>     aws iam attach-role-policy --role-name aurora-monitoring \

Reviewer 2: ```
aws iam attach-role-policy --role-name aurora-monitoring \
```

Author: Updated accordingly.

                           --policy-arn arn:aws:iam::aws:policy/service-role/AmazonRDSEnhancedMonitoringRole

```

>   Create an 2nd IAM role and attach policy to the role (This role is required for aurora proxy to fetch db secrets from secret manager )

Reviewer 2: Why does this need to be done without Crossplane?

```shell

>     aws iam create-role --role-name rds-proxy \

Reviewer 2: ```suggestion
aws iam create-role --role-name rds-proxy \
```

                        --assume-role-policy-document '{    
        "Version": "2012-10-17",
        "Statement": [
        {
            "Sid": "",
            "Effect": "Allow",
                "Principal":
                {
                "Service": "rds.amazonaws.com"
                },
                "Action": "sts:AssumeRole"

        }
    ]
    }'
```

 ```shell

>     aws iam put-role-policy --role-name rds-proxy \

Reviewer 2: ```suggestion
aws iam put-role-policy --role-name rds-proxy \
```

Author: Updated accordingly

                            --policy-name rds-proxy-policy \  
                            --policy-document '{
                            "Version": "2012-10-17",
                            "Statement": [
                                {
                                    "Sid": "getsm",
                                    "Effect": "Allow",
                                    "Action": "secretsmanager:GetSecretValue",
                                    "Resource": "*"
                                },
                                {
                                    "Sid": "kmsdecrypt",
                                    "Effect": "Allow",
                                    "Action": "kms:Decrypt",
                                    "Resource": "*",
                                    "Condition": {
                                        "StringEquals": {

>                                             "kms:ViaService": "secretsmanager.us-east-1.amazonaws.com"

Reviewer 2: Is it guaranteed to be in `us-east-1`?

Author: Thats need to updated accordingly based on the region where we create the RDS.

Author: I will update that accordingly.

Author: updated

                                        }
                                    }
                                }
                            ]
                        }'
  
```
  ### Provide value to the claim before applying.
     
     1. Open the claim and substitute aurora-monitoring role arn with monitoringRoleArn and rds-proxy role arn with proxyRoleArn respectively.
     2. Make sure check the CIDR block for application and provide in the claim. This will used as the security group 

>         ingress rule for proxy from application ( we can also provide the security group id instead of CIDR).    

Reviewer 2: Should be placed outside code block. It's harder to read

Author: updated .

    
 ### Verify your XRDS and composition

  Verify the XRDs

```shell

>     kubectl get xrds

Reviewer 2: ```suggestion
kubectl get xrds |  grep 'xauroras.db.awsblueprint.io'
```

Author: updated accordingly.

```

 Expected output

    NAME                          ESTABLISHED   OFFERED   AGE
    xauroras.db.awsblueprint.io   True          True      5m
    

    Verify the Compositions
    
 ```shell
    

>      kubectl get compositions

Reviewer 2: ```suggestion
kubectl get compositions | grep xauroras.db.awsblueprint.io
```

Author: updated accordingly.

        --namespace team-a \
        --cluster ${CLUSTER_NAME} \
        --attach-policy-arn ${POLICY_ARN} \
        --approve \
        -override-existing-serviceaccounts
``` 
    Generate a DB token which can be used for password when asked
    
```shell
    
    PROXY_TERGET-ENDPOINT=k get secrets aurora-cluster-secrets  -n team-a -o json | jq -r .data.proxyEndpoint | base64 --decode
    CLUSTER_USER_NAME=k get secrets aurora-cluster-secrets  -n team-a -o json | jq -r .data.clusterUsername | base64 --decode
    
    aws rds generate-db-auth-token \
    --hostname ${PROXY_TERGET-ENDPOINT}  \
    --port 5432 \
    --region us-east-1 \
    --username ${CLUSTER_USER_NAME}
    
```
   create the pod with pgsql client:

```shell

    apiVersion: v1
    kind: Pod
    metadata:
    name: postgres-client
    namespace: team-a
    spec:
    serviceAccountName: rds-access
    containers:
    - name: postgreclient
        image: postgres:latest
        command: ["sleep"]
        args: ["3600"]  # Sleep for 1 hour (3600 seconds)
        envFrom:
        - secretRef:
            name: aurora-cluster-secrets
    volumes:
        - name: secret-volume
        secret:

>             secretName: aurora-cluster-secrets # This secret name can be changed in the claim

Reviewer 2: I would rather put this in a file so users can just `k apply -f file`

Author: moved to a different file..

```

   Exec into the pod

```shell

    k exec -it postgres-client -n team-a  -- sh
   
```

   create a DB connection and provide the token which you generate earlier for password.

```shell
   
   psql -h ${PROXY_TERGET-ENDPOINT}  -U ${CLUSTER_USER_NAME} -d aurorapgsqldb -W
```
    Note : we are providing the name of the database in the claim as:  aurorapgsqldb
   

>     Now we should be able to connect to the DB with psql client.

Reviewer 2: Provide basic examples commands like listing tables to prove it works.

Author: Updated accordingly
